### ISSUE TITLE
I am unable to create a simple custom map

### ISSUE THREAD

**Issue by pascualpobil** [bpstahlman/txtfmt#24]
Hi, I'm a pretty noob vim user and I'm trying to create a simple custom map in order to highlight previously selected text (in visual mode). All I want to do is select some text, press something like Leader h and have my selected text highlighted.

Ive been looking for plugins to do this task for a couple days but all I found is temporary or search/pattern related plugins. If I'm not wrong this is the only plugin able to highlight a single arbitrary chunk of text and make the highlight persistent.

I've spent a couple hours on the help doc, but I'm certainly sure I'm not going to be able to get such amount of information together. I think I was kind of close in the User-mappings section, but to be fair it is still chinese for me. I could try to explain what I 'understood', but I don't think it's going to help.

TLDR: im looking for some simple mapping like:
vnoremap Leader h          foregroundcolor = x      backgroundcolor = y

so that i press Leader h and the plugin applies the configured fc and bc to my text (selected in visual mode).
I also need to say that I'd like to use a global Leader key instead of a localleader, if possible.

I know your plugin is massive and I would love to be able to understand it and leverage its power, but I'm pretty limitated.

Thanks in advance.

**Comment by bpstahlman** [comment]
First of all, did you go through the examples in the [Auto Maps](https://github.com/bpstahlman/txtfmt#auto-maps) section of the README.md? Auto maps make it pretty easy to do what you're trying to do. For instance, if you've selected some text and want to highlight it blue, just hit `\h` and enter `kb` at the prompt. The only drawback is that you have to enter the `kb` each time you want to highlight a piece of text blue.

The existing user-maps feature attempts to address this limitation, but because the feature predates auto maps, it's limited to manual maps, which makes what you're trying to do very tedious and error-prone. The good news is I'm close to releasing a _*much*_ better version of user-maps, called [shortcut maps](https://github.com/bpstahlman/txtfmt/issues/11#issue-152036238), which makes it easy and intuitive to do what you're trying to do. Feature development is on the "shortcut-maps" branch. I'm hoping to be ready to release within the next several weeks.

Having said all that... It is _sort of_ possible to do what you're trying to do with the legacy user-maps. Here's how you could map `\b` to give the visually highlighted text a blue background and red foreground:
```
" This global variable needs to be assigned before the plugin loads.
" Note: You can reload the plugin with :Refresh
let g:txtfmtUsermap1 = 'nnoremap <plug>(bg_blue_fg_red) `><<n\va:k-,c->>`<lt><<n\vi:kb,cr>>'
xmap \b <Esc><plug>(bg_blue_fg_red)
```
One drawback is that because it's based on manual maps, it doesn't take the highlighting of the surrounding text into account, so if you're attempting to highlight text within an existing highlighting block, the results may not be what you expect.

**Comment by pascualpobil** [comment]
Yes I did use the automap \h, but I wanted to record a custom map in my init.vim for the automap feature (withouth success), so I'm so happy to hear that you are already making it easier.

Your solution works like magic.  I'm now wondering how I could remove the highlight and I am again at the documentation, in the User-map expansion macros section.

I'm trying to understand the {map_rhs} format. It's being hard for me to visualize what are the fields <<{mode}{cmd} [{cmd-modifiers}] [:{cmd-parameters}]>> from your example.

The first thing that confuses me is the leading `>
although I cant get to interpret what is the rest for either.

Thanks for your help, it is highly appreciated.

**Comment by bpstahlman** [comment]
The leading \`> was simply to position the cursor at the end of the visual selection before appending the "end color/end background" tokens. (Note that it's necessary to insert the trailing tokens first, since once they've been inserted, \`> will no longer point to the end of the originally selected text.) Here's an analysis of that string:

| Sequence | Field | Meaning |
| -----------|-------|-----------|
| \`>     |                                   | Position at end of visual region |
| <<     |                                  | Begin txtfmt token command |
| n       | mode                       | Indicates we expect to be in 'normal' mode when preceding << is encountered |
| \        | cmd                          | Insert token(s) command |
| v        | cmd-modifiers            | Leave cursor in normal mode after inserting token(s) |
| a        | cmd-modifiers            | Use 'append' to insert the token(s) (i.e., after cursor position) |
| :         |                                    | End command modifiers |
| k-,c-   | cmd-parameters        | Specifies "no background" and "no foreground" tokens |
| >>     |                                  | End txtfmt token command |
| `\<lt> |                                 | Jump to beginning of visual region (i.e., \`<) |
| <<     |                                  | Begin txtfmt token command |
| n       | mode                         | Indicates we expect to be in 'normal' mode when preceding << is encountered |
| v      | cmd-modifiers           | Leave cursor in normal mode after inserting token(s) |
| i       | cmd-modifiers           | Use 'insert' to insert the token(s) (i.e., before cursor position) |
| :        |                                    | End command modifiers |
| kb,cr  | cmd-parameters     | Specifies "blue background" and "red foreground" tokens |
| >>     |                                | End txtfmt token command |

Admittedly, this is _much_ too complex. The average user simply should not need to deal with this type of complexity. In fact, this was really part of the motivation for introducing auto maps, and now that auto maps exist, "shortcut maps" are relatively easy to implement.

Incidentally, there's a relatively easy way to create a map to remove highlighting even before the release of "shortcut-maps", but you'd need to comment the following 2 lines in function `s:Prompt_fmt_clr_spec()` (within ftplugin/txtfmt.vim):
```
"call inputsave()
.
.
"call inputrestore()
```
If you did that, you could create a mapping to remove background/foreground colors in the visual selection like this;
```
let maptext= 'gv\hc-,k-' . "\<cr>"
xmap \b <esc>:call feedkeys(maptext)<cr>
```
At any rate, I'll update this thread once "shortcut-maps" have been released...

**Comment by pascualpobil** [comment]
I am really grateful that you provided such detailed explanation. I am going to take my time to analize it and also try to comment those lines and make the deleter command work. I'll come back soon with the results!

Thank you very much. Have a good day sir.

**Comment by bpstahlman** [comment]
No problem. You too!

**Comment by bpstahlman** [comment]
@pascualpobil Although the feature hasn't yet been merged to master or released on vim.org, [the branch that contains it](https://github.com/bpstahlman/txtfmt/tree/shortcut-maps) is pretty stable and [a section has been added](https://github.com/bpstahlman/txtfmt/blob/shortcut-maps/README.md#shortcut-maps) to the README to document the intended usage.

If you get a chance to try it out, I'd definitely appreciate feedback on any aspects of usability, documentation, option defaults, default shortcuts, etc... Nothing is set in stone at this point.

**Comment by pascualpobil** [comment]
@bpstahlman Hi! I've been reading the new docs for a while but it is still kinda tricky for me. I'd love to have more time and knowledge to dive into it, but right now I'm feeling a bit frustrated.

Apart from that, I've been using the solution you gave me and I'm in love with it. I even got to enable all colors in the background. When it came to change one of the colors of the palette I didn't have the same luck tho (the stock blue is not -but almost- the same blue I have in my colorscheme), so I'm just using yellow over black and I'm happy with it. 

I've also come to an error that im going to try to collect, that appears somewhat randomly, and I think it has something to do with a mapping conflict with any other plugin. It says something about <lt> is already mapped. I guess it is something related with internal mechanisms of txtfmt.

I know that vim and this plugin is about possibilities, features and complete control, but in my opinion it would be awesome if this plugin had some 'easy mode', in which you could do something like:

nnnoremap <leader>w/e :txtfmthighlight fc=#000000 bc=#333333

and have it highlight what you select in visual mode (that's the use im giving it right now).

I've been using the plugin only in visual mode and find it perfect, super fast and handy. As I do not understand properly select or operator mode (I guess when you talk about operator you refer to operator-pending) I cannot appreciate the benefits of them.

Changing topic, I've realized that txtfmt marks the text to highlight it. I don't know if this only happens in manual maps or  if the highlighting mechanism is the same among all interfaces (manual maps/user maps/shortcut maps).
I was wondering if it would be possible to do highlighting just like vim does it with folds, using views. I am zero documented about this, but it comes to my mind quite often.

Another thing is that, sometimes, the cursor locks in visual mode and shows some abnormal behaviour when trying to move it over previously highlitghted text. I try to have my text in its definitive position before highlighting it or, if I have done it already, I undo it, I move it and then I highlight it again. It does some other weird things I cannot track nor understand, but I usually just get over it with a little bit of time and luck.

I haven't had much time to read the docs, understand the errors or the cursor behaviour, but for now I'm really satisfied with the little use im being able to give to your plugin and the solution you gave me. I'm sorry I cannot give better quality feedback, but It really takes me time to understand docs at this level and I need to focus on studying very hard right now.

My most honest feedback is again, a minimal interface (nnoremap style, with hex color parameters) with minimal functionality: coloring in visual mode. That would absolutely crack it.

From time to time I spend some time to improve my environment, so I will come back here with more energy (hopefully) next time and give those shortcut maps another try.

Cheers!

**Comment by bpstahlman** [comment]
Pascual,
Responses inline...

> @bpstahlman Hi! I've been reading the new docs for a while but it is still kinda tricky for me. I'd love to have more time and knowledge to dive into it, but right now I'm feeling a bit frustrated.
> 
> Apart from that, I've been using the solution you gave me and I'm in love with it. I even got to enable all colors in the background. When it came to change one of the colors of the palette I didn't have the same luck tho (the stock blue is not -but almost- the same blue I have in my colorscheme), so I'm just using yellow over black and I'm happy with it.
> 
> I've also come to an error that im going to try to collect, that appears somewhat randomly, and I think it has something to do with a mapping conflict with any other plugin. It says something about is already mapped. I guess it is something related with internal mechanisms of txtfmt.

I suspect the error/warning occurs because one of the mappings Txtfmt wants to create results in conflict/ambiguity with respect to one of your existing mappings. For example, Txtfmt map `\h` would conflict with an existing `\h` map, and would be ambiguous with an existing mapping from `\ha`, `\hb`, etc... Txtfmt provides an option that allows you to silence such warnings (or have them displayed as warnings rather than errors), and also to specify whether existing maps should be overwritten by the Txtfmt maps. Here's a snippet of Vim script showing various possibilities...


```vim
" Silence all warnings and create Txtfmt maps even when doing so will overwrite an existing map or create an ambiguous map.
let g:txtfmtMapwarn = 'cC'

" Silence all warnings and create Txtfmt maps that will result in ambiguity but skip Txtfmt maps that would overwrite existing maps.
let g:txtfmtMapwarn = 'c'

" Warn about conflicts only. Don't overwrite any existing maps or create ambiguous ones.
let g:txtfmtMapwarn = 'M'
```

> 
> I know that vim and this plugin is about possibilities, features and complete control, but in my opinion it would be awesome if this plugin had some 'easy mode', in which you could do something like:
> 
> nnnoremap w/e :txtfmthighlight fc=#000000 bc=#333333
> 
> and have it highlight what you select in visual mode (that's the use im giving it right now).

Perhaps I misunderstand what you're proposing, but something like this could be accomplished with the new shortcuts feature simply by having the following in your vimrc:
```
let g:txtfmtShortcuts = [
    'w/e kMyColor'
\ ]
```
...where MyColor has been defined as `#333333` using `txtfmtBgcolor`.
**Note:** I'm assuming the `w/e` in your example was the map lhs, but perhaps it signified something else?


> 
> I've been using the plugin only in visual mode and find it perfect, super fast and handy. As I do not understand properly select or operator mode (I guess when you talk about operator you refer to operator-pending) I cannot appreciate the benefits of them.

Txtfmt operator maps are sometimes more convenient than visual mode because they don't require you to make a visual selection before highlighting. For example, to highlight the current word, you would hit `\hiw` in normal mode, then enter the desired highlighting spec at the prompt. With the default shortcut maps, making the current word bold would be even easier: just hit `,biw` with the cursor anywhere on the word you wish to highlight. Note that in both cases, the `iw` is Vim's "inner word" _text object_. Txtfmt's operators also work with motions. For example, to highlight the current and subsequent 3 lines red, you could use the default shortcut `,R` (red background) in conjunction with Vim's builtin `j` motion (down line) with a count of 3: i.e., in normal mode, hit `_R3j`


Select mode is like visual mode, but typically entered by selecting with the mouse, rather than selecting with `v` (characterwise visual) or `V` (linewise visual). I don't personally use select mode much. The main way it differs from visual mode is that in select mode, typing printable characters causes the selection to be overwritten. In other words, it works more like the selection in a typical Windows application.

> 
> Changing topic, I've realized that txtfmt marks the text to highlight it. I don't know if this only happens in manual maps or if the highlighting mechanism is the same among all interfaces (manual maps/user maps/shortcut maps). I was wondering if it would be possible to do highlighting just like vim does it with folds, using views. I am zero documented about this, but it comes to my mind quite often.

Yes. the same token-based highlighting mechanism is used for both manual maps and auto maps. Vim's `:match` command and `matchadd()` function are nice, but the highlighting they create is not persistent. The fact that the tokens remain part of the Txtfmt file means that the highlighting is unlikely to be lost (as it might be if it were simply saved to some sort of settings file, separate from the Txtfmt file). If you wish to print a buffer containing Txtfmt highlighting, or display it on a system that doesn't have the Txtfmt plugin installed, you can easily create an HTML file that looks identical to the Txtfmt buffer using the `:TOhtml` command that's packaged with Vim.
```
:help :TOhtml
```
> 
> Another thing is that, sometimes, the cursor locks in visual mode and shows some abnormal behaviour when trying to move it over previously highlitghted text. I try to have my text in its definitive position before highlighting it or, if I have done it already, I undo it, I move it and then I highlight it again. It does some other weird things I cannot track nor understand, but I usually just get over it with a little bit of time and luck.

I've never seen anything like this. It doesn't sound like it's directly related to the Txtfmt plugin, though it could have something to do with the way Vim processes syntax regions.

> 
> I haven't had much time to read the docs, understand the errors or the cursor behaviour, but for now I'm really satisfied with the little use im being able to give to your plugin and the solution you gave me. I'm sorry I cannot give better quality feedback, but It really takes me time to understand docs at this level and I need to focus on studying very hard right now.

Understood. I appreciate the feedback you've already provided.

> 
> My most honest feedback is again, a minimal interface (nnoremap style, with hex color parameters) with minimal functionality: coloring in visual mode. That would absolutely crack it.

I have been attempting to make things simpler for the typical user. That goal sometimes conflicts with the goal of ensuring backwards compatibility. An "easy mode" is not a bad idea. I'll think some more on that...

> 
> From time to time I spend some time to improve my environment, so I will come back here with more energy (hopefully) next time and give those shortcut maps another try.

Sounds good. If you do, I think you'll find that they're actually much easier to use than the original user maps, and much more convenient than re-entering the same highlighting specs again and again.

Thanks for all your feedback.
Brett S.
> 
> Cheers!

**Comment by bpstahlman** [comment]
> I know that vim and this plugin is about possibilities, features and complete control, but in my opinion it would be awesome if this plugin had some 'easy mode', in which you could do something like:
> 
> nnnoremap w/e :txtfmthighlight fc=#000000 bc=#333333
> 
> and have it highlight what you select in visual mode (that's the use im giving it right now).

@pascualpobil 
Sorry. It just occurred to me earlier this morning what you were trying to do. (I guess I had seen but not really noticed the `n` in `nnoremap`... ;)

Anyways, here's how you could do that with shortcut maps...
```vim
" Enable background color 1 (disabled by default)
let txtfmtBgcolormask = '11101000'
" Define custom foreground and background colors.
let txtfmtBgcolor{1} = '^\\%(w\\%[hite]\\)$,g:#333333'
let txtfmtColor{1} = '^\\%(b\\%[lack]\\)$,g:#000000'
" Define a Visual Txtfmt shortcut that applies the custom colors to visual selection.
let txtfmtShortcuts = [
	\ 'v:<plug>(txtfmt-color-word) cblack kwhite'
\ ]
" Define a normal mode map that selects word under or after the cursor and applies the Txtfmt shortcut.
nmap \w eviw<plug>(txtfmt-color-word)

```

**Explanation:** The `txtfmtBgcolor1` and `txtfmtColor1` definitions replace the default color 1 (black) with your custom gui colors. Note that I specified a color mask with `txtfmtBgcolormask` only because the default mask disables all but red, green and blue background colors. If you replaced one of them with the custom color, you wouldn't need to modify the mask. Also note that the names I chose (white and black) were purely arbitrary: you can use whatever names you like when customizing colors.

I then defined a Txtfmt visual shortcut that uses the special `<plug>` mapping to avoid consuming a map key sequence for something that may never be typed. Of course, if the shortcut is something you would use for highlighting generally, you could use a typable key sequence: e.g., instead of `<plug>(txtfmt-color-word)`, something like `\h`.

Finally, I used nmap to define a mapping for `\w` that jumps to the end of the next word (`e`), selects the word with visual mode (`viw`) and invokes the visual Txtfmt shortcut map to perform the highlighting. Note that I used nmap, rather than nnoremap, to allow the Txtfmt map to be triggered. Also note that I used `e` to jump to the end of the word before making the visual selection because `viw` will select space between words if the cursor is not on a word.

I've verified that with these definitions, hitting `\w` anywhere on a word (or in the space before it) causes the custom foreground and background colors to be applied.

Sorry for any confusion caused by my earlier response...
Brett S.

**Comment by bpstahlman** [comment]
@pascualpobil The shortcut-maps feature has been merged to master and will be uploaded to the Vim scripts site as version 5.0 within the next day or so...
